src/NetMQ/NetMQPoller.cs

```diff
@@ -243,33 +275,18 @@ public void Remove(ISocketPollable socket)
                 socket.Socket.EventsChanged -= OnSocketEventsChanged;
                 m_sockets.Remove(socket.Socket);
                 m_isPollSetDirty = true;
-            });
+            })
+            .Wait(); 
+            // keep the API syncronous by blocking the calling thread via Wait(), else RemoveThrowsIfSocketAlreadyDisposed() test fails
         }
 
         public void RemoveAndDispose<T>(T socket) where T : ISocketPollable, IDisposable
         {
-            if (socket == null)
-                throw new ArgumentNullException(nameof(socket));
-            if (socket.IsDisposed)
-                throw new ArgumentException("Must not be disposed.", nameof(socket));
-            CheckDisposed();
-
-            Run(() =>
-            {
-                // Ensure the socket wasn't disposed while this code was waiting to be run on the poller thread
-                if (socket.IsDisposed)
-                    throw new InvalidOperationException(
-                        $"{nameof(NetMQPoller)}.{nameof(RemoveAndDispose)} was called from a non-poller thread, " +
-                        "so ran asynchronously. " +
-                        $"The {socket.GetType().Name} being removed was disposed while the remove " +
-                        $"operation waited to start on the poller thread. When using {nameof(RemoveAndDispose)} " +
-                        "you should not dispose the pollable object .");
+            //call the remove method
+            Remove(socket);
 
-                socket.Socket.EventsChanged -= OnSocketEventsChanged;
-                m_sockets.Remove(socket.Socket);
-                m_isPollSetDirty = true;
-                socket.Dispose();
-            });
+            //dispose of socket
+            socket.Dispose();
```

Comment:
This refactoring only works because of the synchronous wait, which isn't something we can safely do unfortunately.

If you add a `RemoveAndDisposeAsync` overload, its implementation could be simpler by awaiting `RemoveAsync` before disposing.

Reply:
In latest.